 ___
# Atividade: Modelos probabilísticos discretos
___

## Aula 12

**Referência de Leitura:**
1. Magalhães e Lima (7ª. Edição): Seções 3.2 e 3.3

**Hoje:**
1.	Descrever modelos discretos quanto aos resultados teóricos.
2.	Explicar a utilização de modelos probabilísticos no contexto da literatura estatística.
3.	Contrastar resultados teóricos e empíricos.
4.	Fechamento do conteúdo.

**Próxima aula:**
1. Leitura prévia necessária: Magalhães e Lima (7ª. Edição): Capítulo 6 - Seção 6.1.

___

In [4]:
%matplotlib inline
import pandas as pd
import matplotlib.pyplot as plt

from scipy import stats #importa apenas as funções de estatísticas da biblioteca SciPy.
import itertools #importante para gerar todos os resultados possíveis!!


___
## <font color='blue'>Exercício 1</font>

### <font color='red'>**Publicar no github até dia 10/04 INDIVIDUAL **</font>


#### Este exercício irá explorar  construção de árvore de probabilidades com eventos e construção da distribuição de probabilidades de uma variável aleatória discreta. Ainda, irá explorar o modelo de distribuição probabilística Binomial

Um resistor de 100K Ohms comprado na Santa Efigênia tem probabilidade de falha de 5%, segundo um fabricante ruim.

Ainda, por suposição, admita que os resistores falhem de forma independente entre si.

Um aluno de engenharia compra um pacote contendo $n$ resistores.

1. Assuma $n=3$ e faça os itens a seguir.
  1. Monte o espaço amostral que combine todos os resultados possíveis de Falha e não Falha desses $n$ resistores. Ainda, calcule a probabilidade de cada resultado do espaço amostral acontecer.
  2. Se $Y$ representa o número de resistores que falham em um pacotinho com $n$ resistores, então construa a distribuição de probabilidades de $Y$.
  3. Calcule a média e o desvio padrão de $Y$. Interprete os resultados.
  4. Utilizando o modelo probabilístico da distribuição Binomial, encontre a distribuição de probabilidades de cada possível valor de $Y$. Calcule também $E(Y)$ e $Var(Y)$. Compare com as distribuições de probabilidades obtidas nos itens anteriores e discuta porque esses valores são iguais ou diferentes!
   
2. Assuma $n=7$ e repita os itens A) a D) anteriores. No caso de $n=7$ ou maiores quantidades de resistores em um pacotinho, qual a viabilidade de construir a distribuição de probabilidades do número de resistores com defeitos quando $n$ não é pequeno seguindo os itens de A) a C)?
 
3. Com $n=100$, por exemplo, é extremamente inviável a construção de árvore de probabilidades até mesmo utilizando a poderosa ferramenta Python. 
  1. No caso assuma $n=100$ e repita o item 1.D) anterior. 
  2. Ainda, calcule a probabilidade de, em um pacote com 100 resistores, pelo menos a metade deles apresentarem falha?


A seguir, um resumo sobre o modelo teórico Binomial!    
> 

> **DISTRIBUIÇÃO BINOMIAL - TEORIA:**

> É responsável por modelar uma variável $Y$ definida pelo Número de sucessos em $n$ tentativas, cuja probabilidade de sucesso é dada por $p$. Assim, dizemos que $Y$~$Bin(n,p)$ e lê-se: $Y$ segue uma distribuição binomial com $n$ tentativas e $p$ como probabilidade de sucesso. 

> A função de probabilidade nesse caso é dada por:
$P(Y=y)=\left(
\begin{array}{c}
    n \\
    y
\end{array}\right) p^y (1-p)^{(n-y)}$

> Para um experimento ter as probabilidade dos números de sucessos modeladas por uma distribuição Binomial, o mesmo precisa ter as seguintes propriedades:
* é uma contagem de $n$ repetições (ou tentativas ou ensaios) idênticas;
* cada repetição tem apenas 2 resultados possíveis: um é denominado sucesso e o outro, fracasso;
* a probabilidade de sucesso para cada ensaio é denominada $p$ e será constante em cada repetição. Então, a probabilidade de fracasso $(1-p)$ também não varia de tentativa para tentativa;
* As tentativas são independentes.


> **DISTRIBUIÇÃO BINOMIAL - PYTHON:** <sup> [https://en.wikipedia.org/wiki/SciPy]

> Algumas funções úteis para desenvolver seu código: <sup> [http://docs.scipy.org/doc/scipy-0.16.1/reference/generated/scipy.stats.binom.html]
* $P(Y=y)$: `stats.binom.pmf(y, n, p)`
* $P(Y\leq y)$: `stats.binom.cdf(y, n, p)`
* $E(Y)$ e $Var(Y)$: `stats.binom.stats(n, p, moments='mv')`



In [36]:
#item 1

n=3
        
sf = ["S", "F"]
probabilidade = {"S": 0.95, "F":0.05}
itertooing = itertools.product(sf, repeat=n)
e_a = list(itertooing)

probabilidades = []

for ser in e_a:
    p = 1.0
    for e in ser:
        p*=probabilidade[e]
    probabilidades.append(p)
    print("Probabilidade de acontecer ",ser," é igual a ","%3.5f"%p)          



Probabilidade de acontecer  ('S', 'S', 'S')  é igual a  0.85737
Probabilidade de acontecer  ('S', 'S', 'F')  é igual a  0.04512
Probabilidade de acontecer  ('S', 'F', 'S')  é igual a  0.04512
Probabilidade de acontecer  ('S', 'F', 'F')  é igual a  0.00238
Probabilidade de acontecer  ('F', 'S', 'S')  é igual a  0.04512
Probabilidade de acontecer  ('F', 'S', 'F')  é igual a  0.00238
Probabilidade de acontecer  ('F', 'F', 'S')  é igual a  0.00238
Probabilidade de acontecer  ('F', 'F', 'F')  é igual a  0.00013


In [57]:
arex = []

for alex in range(0,3):
    statsb = stats.binom.pmf(alex, 3, 0.05)
    arex.append(statsb)
    print(statsb)

0.857375
0.135375
0.007125000000000007


In [44]:
mediam = 3*arex[0]+2*arex[1]+1*arex[2]
mediam

2.8499999999999996

In [51]:
statsb_2 = stats.binom.stats(3, 0.05, moments='mv')
statsb_2

(array(0.15), array(0.1425))

In [53]:
print("Média dos resistores que não funcionaram -- 0.15 e a Variância -- 0.1425")

dp = 0.1425**(1/2)

print("Desvio padrão = {0}".format(dp))

Média dos resistores que não funcionaram -- 0.15 e a Variância -- 0.1425
Desvio padrão = 0.37749172176353746


In [59]:
#item 2

n=7
        
sf = ["S", "F"]
probabilidade = {"S": 0.95, "F":0.05}
itertooing = itertools.product(sf, repeat=n)
e_a = list(itertooing)

probabilidades = []

for ser in e_a:
    p = 1.0
    for i in ser:
        p*=probabilidade[i]
    probabilidades.append(p)
    print("Probabilidade de acontecer ",ser," é igual a ","%3.5f"%p)     

Probabilidade de acontecer  ('S', 'S', 'S', 'S', 'S', 'S', 'S')  é igual a  0.69834
Probabilidade de acontecer  ('S', 'S', 'S', 'S', 'S', 'S', 'F')  é igual a  0.03675
Probabilidade de acontecer  ('S', 'S', 'S', 'S', 'S', 'F', 'S')  é igual a  0.03675
Probabilidade de acontecer  ('S', 'S', 'S', 'S', 'S', 'F', 'F')  é igual a  0.00193
Probabilidade de acontecer  ('S', 'S', 'S', 'S', 'F', 'S', 'S')  é igual a  0.03675
Probabilidade de acontecer  ('S', 'S', 'S', 'S', 'F', 'S', 'F')  é igual a  0.00193
Probabilidade de acontecer  ('S', 'S', 'S', 'S', 'F', 'F', 'S')  é igual a  0.00193
Probabilidade de acontecer  ('S', 'S', 'S', 'S', 'F', 'F', 'F')  é igual a  0.00010
Probabilidade de acontecer  ('S', 'S', 'S', 'F', 'S', 'S', 'S')  é igual a  0.03675
Probabilidade de acontecer  ('S', 'S', 'S', 'F', 'S', 'S', 'F')  é igual a  0.00193
Probabilidade de acontecer  ('S', 'S', 'S', 'F', 'S', 'F', 'S')  é igual a  0.00193
Probabilidade de acontecer  ('S', 'S', 'S', 'F', 'S', 'F', 'F')  é igual a  

In [56]:
arex = []

for alex in range(0,7):
    statsb = stats.binom.pmf(alex, 7, 0.05)
    arex.append(statsb)
    print(statsb)

0.69833729609375
0.2572821617187501
0.04062349921875004
0.00356346484375
0.00018755078124999986
5.922656250000019e-06
1.0390624999999968e-07


In [65]:
#item 3

aga = 0
stats_3 = stats.binom.stats(100, 0.05, moments='mv')
print(stats_3)

for i in range(0,26):
    stats_4 = stats.binom.pmf(i, 100, 0.05)
    aga = aga + stats_4
    print(stats_4)
print(aga)

(array(5.), array(4.75))
0.005920529220334023
0.031160680107020593
0.08118177185776551
0.1395756779308903
0.17814264156968956
0.18001782727043672
0.15001485605869255
0.10602553736478391
0.06487088799292766
0.034901296464031804
0.016715884095931225
0.007198227601119233
0.0028098344583314945
0.001001074624830605
0.00032741914421153287
9.880016281470898e-05
2.762504552384916e-05
7.184222365335237e-06
1.7435393459730753e-06
3.960394082265668e-07
8.441892649039948e-08
1.6926100549455788e-08
3.1989520177198604e-09
5.709799940094749e-10
9.641548144458742e-11
1.5426477031134177e-11
0.999999999997256


Média --- 5

Variância --- 4.75

Probabilidade de 0 falharem --- 0.6%

Prob 50 em 100 falharem --- 0%


___
## <font color='blue'>Exercício 2</font>

Esse exercício irá explorar modelos de distribuição de Poisson.

Uma fábrica produz determinado veículo. O gerente de vendas fechou um negócio que implica na produção de pelo menos 110 unidades nos próximos dois meses. Para isso, será construída uma nova linha de montagem com capacidade para produzir em média 50 veículos por mês (considere meses de 30 dias). A linha funcionará ininterruptamente. Além disso, apenas um veículo é produzido por vez. Considere que um modelo Poisson seja adequado para a quantidade de veículos produzidos.
   1. Qual a probabilidade  da produção ser igual a 100 veículos nos próximos dois meses?
   2. Qual a probabilidade  da produção ser de pelo menos 100 veículos nos próximos dois meses?
   3. Qual a probabilidade da produção ser entre 100 e 110 veículos nos próximos dois meses?
   4. Se forem produzidos de pelo menos 100 veículos no próximo mês, qual a probabilidade do gerente cumprir a meta negociada?
   5. Caso o prazo seja cumprido, o lucro será da ordem de $250$ mil, caso não seja, o lucro cai para $200$ mil. Qual é o lucro esperado e seu desvio-padrão?

**Observação:** Explore os comandos de `stats.poisson`. <sup>[http://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.poisson.html]
